Add time to target attainment to final exposure metrics

dominic-irx opened the pull request:
Adds time to target attainment calculation for AUC12, AUC24, and troughs.

jasmineirx commented on R/exposure_metrics.R:
```diff
@@ -132,3 +132,71 @@ calc_auc_from_regimen <- function(
+    terminal_interval = 24,
+    ...
+) {
+  target_type <- match.arg(tolower(target_design$type), mipd_target_types())
```
from a user perspective, it might be helpful to check that `target_type` is one supported by this function. At present, we are only returning NA and so it isn't obvious to the user that the reason we are returning NA is due to the target_type simply not being a supported target (versus all their patients failing to attain target)

(the current supported types are any type of AUC, plus trough/cmin)

we currently only use this function for supported targets (except cum_auc i think?, we could write around that or extend it), so I think it would be okay to throw a hard `stop()` for unsupported target types.

if we needed this function to always return a valid number, we could return NA_real_ for unsupported numbers and Inf for never-gets-to-target.

jasmineirx commented on R/exposure_metrics.R:
```diff
@@ -132,3 +132,71 @@ calc_auc_from_regimen <- function(
+) {
+  target_type <- match.arg(tolower(target_design$type), mipd_target_types())
+
+  iov <- PKPDsim::get_model_iov(model)
+  if (is.null(iov[["bins"]])) iov[["bins"]] <- c(0, 9999)
```
is this intended to be passed to PKPDsim::sim a couple lines below? if so, this step was omitted.

jasmineirx commented on R/exposure_metrics.R:
```diff
@@ -132,3 +132,71 @@ calc_auc_from_regimen <- function(
   )
 }
+
+#' Calculate time to target attainment
+#'
```
I think it would be helpful to provide more documentation for the user about how time to target is defined. For example, how dose index is defined, that the time involves the infusion length, etc.

jasmineirx commented on R/exposure_metrics.R:
```diff
@@ -132,3 +132,71 @@ calc_auc_from_regimen <- function(
+#' @param target_design target design, created using `create_target_design()`
+#' @param auc_comp compartment to look for AUC
+#' @param model PKPDsim model object
+#' @param terminal_interval duration of final dosing interval
```
This argument requires clearer documentation. Is this something that has to be specified by the user or can it be known based on target_type / regimen?

jasmineirx commented on R/exposure_metrics.R:
```diff
@@ -132,3 +132,71 @@ calc_auc_from_regimen <- function(
+  if (!is.null(dose_idx)) {
+    regimen$dose_times[dose_idx] + regimen$t_inf[dose_idx]
+  } else {
+    NA_real_
```
Not because I think it is _wrong_ but more out of curiosity: why did you settle on NA_real_ as a return value if the patient never achieves target, versus e.g., Inf?

dominic-irx replied:
Good point. I like this flow:

if not enough data, return NA_real_
if dose identified, return end of infusion
if no dose found, return Inf

jasmineirx commented on R/run_trial.R:
```diff
@@ -125,9 +125,22 @@ run_trial <- function(
         target_design = design$target,
         covariates = covs
       )
+      time_to_target <- calc_time_to_target(
+        regimen = res$final_regimen,
+        target_design = design$target,
+        auc_comp = attr(design$sim$model, "size"),
+        ode = design$sim$model,
+        parameters = pars_true_i,
+        covariates = covs
+      )
```
terminal_interval is not defined here. should it be?

jasmineirx commented on tests/testthat/test-exposure_metrics.R:
```diff
@@ -169,3 +169,104 @@ test_that("handles IOV correctly", {
+    covariates = NULL,
+    parameters = list(CL = 5, V = 50)
+  )
+  expect_equal(time_to_target_trough, 13)
```
```suggestion
  # attains target on second dosing interval + 1-hr infusion = 13 hr
  expect_equal(time_to_target_trough, 13)
```

very optional suggest but it might be helpful to have some code comments in the test file to explain why we expect a value of e.g. 13 here.

jasmineirx replied:
My comment above also applied to the other unit test expectations, if you could add some comments there too?

jasmineirx commented on tests/testthat/test-exposure_metrics.R:
```diff
@@ -169,3 +169,104 @@ test_that("handles IOV correctly", {
+  expect_equal(time_to_target_auc12, 13)
+})
+
+test_that("calc_time_to_target: return NA when unsupported type", {
```
see above for suggested change in behavior regarding this case.

jasmineirx commented on tests/testthat/test-exposure_metrics.R:
```diff
@@ -169,3 +169,104 @@ test_that("handles IOV correctly", {
   expect_false(result2a == result2b) # iov is different after day 3
 })
 
+test_that("calc_time_to_target: correct time to target calculated", {
```
it would be good to add a test for when a supported target is used but the encounter never attains target.

jasmineirx reviewed (commented):
a few comments. overall looks good, most of my suggestions are about making it more user-friendly

dominic-irx commented:
Changes:

- removed terminal interval as an argument; instead it is the dosing interval between the last two doses for trough target and 12 or 24 hours for AUC12 or 24 targets
- added a check for target type, now returns NA_real_ if target is not supported
- returns Inf if target is supported but never reached
- added test for Inf case

jasmineirx commented on R/exposure_metrics.R:
```diff
@@ -132,3 +132,102 @@ calc_auc_from_regimen <- function(
+#' @param auc_comp compartment to look for AUC
+#' @param model PKPDsim model object
+#' @param ... arguments passed on to `PKPDsim::sim`
+#' @return a numeric value indicating the time, in hours, if a dosing interval
+#'   contained the target metric within range. Otherwise, returns NA.
```
There are some code comments and function documentation that needs to be updated to reflect the fact that we now return a number, NA or Inf.

jasmineirx commented on tests/testthat/test-exposure_metrics.R:
```diff
@@ -169,3 +169,131 @@ test_that("handles IOV correctly", {
+    covariates = NULL,
+    parameters = list(CL = 5, V = 50)
+  )
+  expect_equal(is.na(time_to_target_unsupported), TRUE)
```
```suggestion
  expect_true(is.na(time_to_target_unsupported))
```
just in case you weren't aware of this one, it is a little more streamlined to write and more expressive for expectations of this type

dominic-irx replied:
I've definitely seen this before, but just forgot here. Thanks for the reminder!

jasmineirx commented on tests/testthat/test-exposure_metrics.R:
```diff
@@ -169,3 +169,131 @@ test_that("handles IOV correctly", {
+    type = "infusion"
+  )
+  unsupported_target <- create_target_design(
+    targettype = "cum_auc",
```
```suggestion
    targettype = "something_unsupported",
```
this is a little more future-proof since then we don't have to change this test when we extend the function.

jasmineirx reviewed (changes requested):
minor requests for documentation only.

it would be nice to add a vignette, but it's fine by me to include that as part of a new PR.

dominic-irx commented:
> minor requests for documentation only.
> 
> it would be nice to add a vignette, but it's fine by me to include that as part of a new PR.

Yeah I added target metrics in the other PR to pegasp, but I'll a separate PR for adding to vignettes

jasmineirx reviewed (approved):
regarding a vignette to show this new capability, rather than adding it to existing vignettes I think it would be nice to have it as a dedicated vignette. e.g., how does MIPD get you to target faster?
